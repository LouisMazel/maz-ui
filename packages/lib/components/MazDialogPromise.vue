<template>
  <MazDialog
    :model-value="currentModal?.isActive ?? false"
    @update:model-value="rejectDialog(currentModal)"
  >
    <template #title>
      <slot name="title">{{ data?.title }}</slot>
    </template>
    <template #default>
      <!--
        @slot Default slot - Place your content
          @binding {Function} resolve resolve function
          @binding {Function} reject reject function
      -->
      <slot
        :resolve="(reason?: string | boolean) => resolveDialog(currentModal, reason)"
        :reject="(reason?: string | boolean) => rejectDialog(currentModal, reason)"
      >
        {{ data?.message }}</slot
      >
    </template>
    <template #footer>
      <!--
        @slot Footer Buttons slot
          @binding {Function} resolve resolve function
          @binding {Function} reject reject function
      -->
      <slot
        :resolve="(reason?: string | boolean) => resolveDialog(currentModal, reason)"
        :reject="(reason?: string | boolean) => rejectDialog(currentModal, reason)"
        name="footer-button"
      >
        <div class="maz-space-x-2">
          <template v-if="buttons.length > 0">
            <MazBtn
              v-for="(button, i) in buttons"
              :key="i"
              :color="button.color"
              :size="button.size"
              :outline="button.outline"
              :rounded="button.rounded"
              :disabled="button.disabled"
              :block="button.block"
              :loading="button.loading"
              @click="
                button.type === 'resolve'
                  ? resolveDialog(currentModal, button.response)
                  : rejectDialog(currentModal, button.response)
              "
            >
              {{ button.text }}
            </MazBtn>
          </template>
          <template v-else>
            <MazBtn color="danger" outline @click="rejectDialog(currentModal)">
              <slot name="cancel-text"> {{ data?.cancelText || 'Cancel' }} </slot>
            </MazBtn>
            <MazBtn color="success" @click="resolveDialog(currentModal)">
              <slot name="confirm-text"> {{ data?.confirmText || 'Confirm' }} </slot>
            </MazBtn>
          </template>
        </div>
      </slot>
    </template>
  </MazDialog>
</template>

<script lang="ts">
  export {
    useMazDialogPromise,
    type DialogState,
    type DialogData,
  } from './MazDialogPromise/use-maz-dialog-promise'
  export type { Color, Size } from './types'
</script>

<script lang="ts" setup>
  import { type PropType, computed, defineAsyncComponent } from 'vue'
  import { type Color, type Size } from './types'
  import {
    useMazDialogPromise,
    type DialogData,
    type DialogState,
  } from './MazDialogPromise/use-maz-dialog-promise'

  export type DialogButton = {
    text: string
    type: 'resolve' | 'reject'
    block?: boolean
    color?: Color
    disabled?: boolean
    loading?: boolean
    outline?: boolean
    response?: string | boolean
    rounded?: boolean
    size?: Size
  }

  import MazDialog from './MazDialog.vue'
  const MazBtn = defineAsyncComponent(() => import('./MazBtn.vue'))

  const props = defineProps({
    /** Dialog Data - type DialogData */
    data: { type: Object as PropType<DialogData>, default: undefined },
    /** Uniq identifier */
    identifier: { type: String, required: true },
    /** Custom buttons - type DialogButton[] */
    buttons: { type: Array as PropType<DialogButton[]>, default: () => [] },
  })

  const { dialogState, rejectDialog, resolveDialog } = useMazDialogPromise()

  const currentModal = computed(
    () => dialogState.value.find(({ id }) => id === props.identifier) as DialogState,
  )
</script>
