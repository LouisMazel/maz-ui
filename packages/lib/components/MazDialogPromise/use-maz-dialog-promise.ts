import { ref } from 'vue'

export interface DialogState {
  id: string
  isActive: boolean
  resolve: (value: unknown) => void
  reject?: (reason?: unknown) => void
}

export interface DialogData {
  title: string
  message: string
  cancelText?: string
  confirmText?: string
}

const confirmDialogData = ref<DialogData>()

const dialogState = ref<DialogState[]>([])

const showDialogAndWaitChoice = (identifier: string, callback?: () => unknown) => {
  return new Promise((resolve, reject) => {
    dialogState.value = [
      ...dialogState.value,
      {
        id: identifier,
        isActive: true,
        resolve: async () => {
          await callback?.()
          resolve(true)
        },
        reject,
      },
    ]
  })
}

const removeDialogFromState = (identifier: string) => {
  dialogState.value = dialogState.value.filter(({ id }) => id !== identifier)
  return dialogState.value
}

const responseDialog = (
  type: 'resolve' | 'reject',
  currentDialog: DialogState,
  response: string | boolean = false,
) => {
  if (currentDialog) {
    currentDialog[type]?.(response)
    currentDialog.isActive = false

    setTimeout(() => {
      removeDialogFromState(currentDialog.id)
    }, 500)
  }
}

export const useMazDialogPromise = () => ({
  confirmDialogData,
  dialogState,
  showDialogAndWaitChoice,
  removeDialogFromState,
  rejectDialog: (currentDialog: DialogState, response: string | boolean = 'cancel') =>
    responseDialog('reject', currentDialog, response),
  resolveDialog: (currentDialog: DialogState, response: string | boolean = 'accept') =>
    responseDialog('resolve', currentDialog, response),
})
