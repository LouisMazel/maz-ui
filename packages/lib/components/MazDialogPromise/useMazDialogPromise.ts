import type { Ref } from 'vue'
import type { Props } from '../MazBtn.vue'
import type { Size } from './../types'
import { ref } from 'vue'

export interface DialogState {
  id: string
  isActive: boolean
  resolve: (value: unknown) => void
  reject?: (reason?: unknown) => void
}

export interface DialogButton extends Props {
  text?: string
  size?: Size
}

export interface ActionButton {
  text: string
  action: () => unknown
}

export interface PromiseButton {
  text: string
  type: 'resolve' | 'reject'
  response?: unknown
}

export type DialogCustomButton = Omit<DialogButton, 'type'> & (PromiseButton | ActionButton)

export interface DialogData {
  /**
   * Dialog title
   */
  title?: string
  /**
   * Dialog message
   */
  message?: string
  /**
   * Dialog cancel text
   * @default 'Cancel'
   */
  cancelText?: string
  /**
   * Dialog cancel button
   * @default { text: 'Cancel', color: 'danger', outline: true }`
   */
  cancelButton?: false | DialogButton
  /**
   * Dialog confirm text
   * @default 'Confirm'
   */
  confirmText?: string
  /**
   * Dialog confirm button
   * @default { text: 'Confirm', color: 'success' }
   */
  confirmButton?: false | DialogButton
  /**
   * This is a list of custom buttons that will replace the default confirm and cancel buttons
   */
  buttons?: DialogCustomButton[]
}

export const defaultData = {
  cancelText: 'Cancel',
  confirmText: 'Confirm',
  cancelButton: {
    text: 'Cancel',
    color: 'danger',
  },
  confirmButton: {
    text: 'Confirm',
    color: 'success',
  },
} satisfies DialogData

const data = ref(defaultData) as Ref<DialogData>

const dialogState = ref<DialogState[]>([]) as Ref<DialogState[]>

function showDialogAndWaitChoice(identifier: string, callback?: () => unknown) {
  return new Promise((resolve, reject) => {
    dialogState.value = [
      ...dialogState.value,
      {
        id: identifier,
        isActive: true,
        resolve: async (response: unknown) => {
          resolve(response)
          await callback?.()
        },
        reject: async (response: unknown) => {
          reject(response)
          await callback?.()
        },
      },
    ]
  })
}

function removeDialogFromState(identifier: string) {
  dialogState.value = dialogState.value.filter(({ id }) => id !== identifier)
  return dialogState.value
}

function responseDialog(type: 'resolve' | 'reject', currentDialog: DialogState, response: unknown) {
  if (!currentDialog) {
    return
  }

  currentDialog[type]?.(response)
  currentDialog.isActive = false

  setTimeout(() => {
    removeDialogFromState(currentDialog.id)
  }, 500)
}

export function useMazDialogPromise() {
  return {
    data,
    dialogState,
    showDialogAndWaitChoice,
    removeDialogFromState,
    rejectDialog: async (currentDialog: DialogState, response: unknown = new Error('cancel'), action?: () => unknown) => {
      await action?.()
      return responseDialog('reject', currentDialog, response)
    },
    resolveDialog: async (currentDialog: DialogState, response: unknown = 'accept', action?: () => unknown) => {
      await action?.()
      return responseDialog('resolve', currentDialog, response)
    },
  }
}
