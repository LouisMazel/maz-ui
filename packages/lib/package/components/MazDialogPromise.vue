<template>
  <MazDialog
    v-if="currentModal"
    v-bind="$attrs"
    :model-value="currentModal?.isActive"
    @update:model-value="rejectDialog(currentModal)"
  >
    <template #title>
      <slot name="title">{{ data?.title }}</slot>
    </template>
    <template #default>
      <slot>{{ data?.message }}</slot>
    </template>
    <template #footer>
      <div class="maz-space-x-2">
        <MazBtn color="danger" outline @click="rejectDialog(currentModal)">
          <slot name="confirm-text"> {{ data?.confirmText || 'Cancel' }} </slot>
        </MazBtn>
        <MazBtn color="success" @click="resolveDialog(currentModal)">
          <slot name="cancel-text"> {{ data?.cancelText || 'Confirm' }} </slot>
        </MazBtn>
      </div>
    </template>
  </MazDialog>
</template>

<script lang="ts">
  export { useMazDialogPromise } from './MazDialogPromise/use-maz-dialog-promise'
  export type {
    DialogState,
    DialogData,
  } from './MazDialogPromise/use-maz-dialog-promise'
</script>

<script lang="ts" setup>
  import { PropType, computed } from 'vue'
  import MazDialog from './MazDialog.vue'
  import MazBtn from './MazBtn.vue'
  import {
    useMazDialogPromise,
    DialogData,
    DialogState,
  } from './MazDialogPromise/use-maz-dialog-promise'

  const props = defineProps({
    data: { type: Object as PropType<DialogData>, default: undefined },
    identifier: { type: String, required: true },
  })

  const { dialogState, rejectDialog, resolveDialog } = useMazDialogPromise()

  const currentModal = computed(() => {
    return dialogState.value.find(
      ({ id }) => id === props.identifier,
    ) as DialogState
  })
</script>
