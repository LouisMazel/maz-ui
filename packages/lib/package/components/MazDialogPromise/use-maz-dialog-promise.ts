import { ref } from 'vue'

export interface DialogState {
  id: string
  isActive: boolean
  resolve: (value: unknown) => void
  reject?: (reason?: unknown) => void
}

export interface DialogData {
  title: string
  message: string
  cancelText?: string
  confirmText?: string
}

const confirmDialogData = ref<DialogData>()

const dialogState = ref<DialogState[]>([])

const showDialogAndWaitChoice = (
  identifier: string,
  callback?: () => unknown,
) => {
  return new Promise((resolve, reject) => {
    dialogState.value = [
      ...dialogState.value,
      {
        id: identifier,
        isActive: true,
        resolve: async () => {
          await callback?.()
          resolve(true)
        },
        reject,
      },
    ]
  })
}

const removeDialogFromState = (identifier: string) => {
  dialogState.value = dialogState.value.filter(({ id }) => id !== identifier)
  return dialogState.value
}

const rejectDialog = (currentDialog: DialogState) => {
  if (currentDialog) {
    currentDialog.reject?.(false)
    currentDialog.isActive = false

    setTimeout(() => {
      removeDialogFromState(currentDialog.id)
    }, 500)
  }
}

const resolveDialog = (currentDialog: DialogState) => {
  if (currentDialog) {
    currentDialog.resolve?.(true)
    currentDialog.isActive = false

    setTimeout(() => {
      removeDialogFromState(currentDialog.id)
    }, 500)
  }
}

export const useMazDialogPromise = () => ({
  confirmDialogData,
  dialogState,
  showDialogAndWaitChoice,
  removeDialogFromState,
  rejectDialog,
  resolveDialog,
})
