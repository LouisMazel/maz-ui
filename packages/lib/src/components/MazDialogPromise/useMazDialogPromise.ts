import type { Ref } from 'vue'
import type { MazBtnProps } from '../MazBtn.vue'
import type { MazSize } from '../types'
import { ref } from 'vue'

export interface MazDialogState {
  id: string
  isActive: boolean
  resolve: (value: unknown) => void
  reject?: (reason?: unknown) => void
}

export interface MazDialogButton extends MazBtnProps {
  text?: string
  size?: MazSize
}

export interface MazDialogActionButton {
  text: string
  onClick: () => unknown
}

export interface MazDialogPromiseButton {
  text: string
  type: 'resolve' | 'reject'
  response?: unknown
}

export type MazDialogCustomButton = Omit<MazDialogButton, 'type'> & (MazDialogPromiseButton | MazDialogActionButton)

export interface MazDialogData {
  /**
   * Dialog title
   */
  title?: string
  /**
   * Dialog message
   */
  message?: string
  /**
   * Dialog cancel text
   * @default 'Cancel'
   */
  cancelText?: string
  /**
   * Dialog cancel button
   * @default { text: 'Cancel', color: 'destructive', outlined: true }`
   */
  cancelButton?: false | MazDialogButton
  /**
   * Dialog confirm text
   * @default 'Confirm'
   */
  confirmText?: string
  /**
   * Dialog confirm button
   * @default { text: 'Confirm', color: 'success' }
   */
  confirmButton?: false | MazDialogButton
  /**
   * This is a list of custom buttons that will replace the default confirm and cancel buttons
   */
  buttons?: MazDialogCustomButton[]
}

export const defaultData = {
  cancelText: 'Cancel',
  confirmText: 'Confirm',
  cancelButton: {
    text: 'Cancel',
    color: 'destructive',
  },
  confirmButton: {
    text: 'Confirm',
    color: 'success',
  },
} satisfies MazDialogData

const data = ref(defaultData) as Ref<MazDialogData>

const dialogState = ref<MazDialogState[]>([]) as Ref<MazDialogState[]>

function showDialogAndWaitChoice(identifier: string, callback?: () => unknown) {
  return new Promise((resolve, reject) => {
    dialogState.value = [
      ...dialogState.value,
      {
        id: identifier,
        isActive: true,
        resolve: async (response: unknown) => {
          resolve(response)
          await callback?.()
        },
        reject: async (response: unknown) => {
          reject(response)
          await callback?.()
        },
      },
    ]
  })
}

function removeDialogFromState(identifier: string) {
  dialogState.value = dialogState.value.filter(({ id }) => id !== identifier)
  return dialogState.value
}

function responseDialog(type: 'resolve' | 'reject', currentDialog: MazDialogState, response: unknown) {
  if (!currentDialog) {
    return
  }

  currentDialog[type]?.(response)
  currentDialog.isActive = false

  setTimeout(() => {
    removeDialogFromState(currentDialog.id)
  }, 500)
}

export function useMazDialogPromise() {
  return {
    data,
    dialogState,
    showDialogAndWaitChoice,
    removeDialogFromState,
    rejectDialog: async (currentDialog: MazDialogState, response: unknown = new Error('cancel'), onClick?: () => unknown) => {
      await onClick?.()
      return responseDialog('reject', currentDialog, response)
    },
    resolveDialog: async (currentDialog: MazDialogState, response: unknown = 'accept', onClick?: () => unknown) => {
      await onClick?.()
      return responseDialog('resolve', currentDialog, response)
    },
  }
}
