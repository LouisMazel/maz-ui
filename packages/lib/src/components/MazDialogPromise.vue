<script lang="ts">
/* eslint-disable import/first */
export { useMazDialogPromise } from './MazDialogPromise/useMazDialogPromise'
export type * from './MazDialogPromise/useMazDialogPromise'
</script>

<script lang="ts" setup>
import type { ComponentPublicInstance } from 'vue'
import type {
  MazDialogCustomButton,
  MazDialogData,
  MazDialogPromiseButton,
  MazDialogState,
} from './MazDialogPromise/useMazDialogPromise'
import { computed, ref } from 'vue'
import MazBtn from './MazBtn.vue'

import MazDialog, { type MazDialogProps } from './MazDialog.vue'

import { defaultData, useMazDialogPromise } from './MazDialogPromise/useMazDialogPromise'

export interface MazDialogPromiseInternalProps {
  /** Dialog Data - @type DialogData */
  data?: MazDialogData
  /** Message to display */
  message?: string
  /** Uniq identifier */
  identifier: string
  /** Custom buttons - @type DialogCustomButton[] */
  buttons?: MazDialogCustomButton[]
}

export type MazDialogPromiseProps = MazDialogPromiseInternalProps & MazDialogProps & MazDialogData

const props = withDefaults(defineProps<MazDialogPromiseProps>(), {
  identifier: undefined,
  message: undefined,
  data: undefined,
  buttons: undefined,
})

defineEmits<{
  /** emitted when modal is open */
  open: [value: void]
  /** emitted when modal is close */
  close: [value: void]
}>()

const { dialogState, rejectDialog, resolveDialog, data: composableData } = useMazDialogPromise()

const customButtons = computed(() => props.buttons ?? props.data?.buttons ?? composableData.value.buttons)

const currentData = computed<MazDialogData>(() => ({
  ...defaultData,
  ...composableData.value,
  ...props.data,
}))

const cancelButtonData = computed(() => {
  const hasButton
      = composableData.value?.cancelButton ?? props.data?.cancelButton ?? defaultData.cancelButton

  if (!hasButton)
    return

  const mergedData = {
    ...defaultData.cancelButton,
    ...composableData.value?.cancelButton,
    ...props.data?.cancelButton,
  }

  return {
    ...mergedData,
    text: props.cancelText || currentData.value.cancelText || mergedData.text,
  }
})
const confirmButtonData = computed(() => {
  const hasButton
      = composableData.value?.confirmButton ?? props.data?.confirmButton ?? defaultData.confirmButton

  if (!hasButton)
    return

  const mergedData = {
    ...defaultData.confirmButton,
    ...composableData.value?.confirmButton,
    ...props.data?.confirmButton,
  }

  return {
    ...mergedData,
    text: props.confirmText || currentData.value.confirmText || mergedData.text,
  }
})

const currentModal = computed(
  () => dialogState.value.find(({ id }) => id === props.identifier) as MazDialogState,
)

const dialog = ref<ComponentPublicInstance<typeof MazDialog>>()

defineExpose({
  close: () => dialog.value?.close?.(),
})

function isPromiseButton(button: MazDialogCustomButton): button is MazDialogPromiseButton {
  return 'type' in button && (button.type === 'resolve' || button.type === 'reject')
}

function customButtonAction(currentModal: MazDialogState, button: MazDialogCustomButton) {
  return isPromiseButton(button)
    ? button.type === 'resolve'
      ? resolveDialog(currentModal, button.response)
      : rejectDialog(currentModal, button.response)
    : resolveDialog(currentModal, undefined, button.action)
}
</script>

<template>
  <MazDialog
    ref="dialog"
    v-bind="{ ...$attrs, ...props }"
    :model-value="currentModal?.isActive ?? modelValue ?? false"
    @close="$emit('close', $event)"
    @open="$emit('open', $event)"
    @update:model-value="rejectDialog(currentModal)"
  >
    <template #title>
      <!--
        @slot title slot - Place your title
      -->
      <slot name="title">
        {{ title || currentData?.title }}
      </slot>
    </template>
    <template #default>
      <!--
        @slot Default slot - Place your content
          @binding {Function} resolve resolve function
          @binding {Function} reject reject function
      -->
      <slot
        :resolve="(reason?: unknown) => resolveDialog(currentModal, reason)"
        :reject="(reason?: unknown) => rejectDialog(currentModal, reason)"
      >
        {{ message || currentData?.message }}
      </slot>
    </template>
    <template #footer>
      <!--
        @slot Footer slot
          @binding {Function} resolve resolve function
          @binding {Function} reject reject function
      -->
      <slot
        :resolve="(reason?: unknown) => resolveDialog(currentModal, reason)"
        :reject="(reason?: unknown) => rejectDialog(currentModal, reason)"
        name="footer-button"
      >
        <div class="maz-flex maz-items-center maz-gap-2">
          <template v-if="customButtons">
            <MazBtn
              v-for="(button, i) in customButtons"
              :key="i"
              v-bind="{
                ...button,
                type: 'button',
              }"
              @click="customButtonAction(currentModal, button)"
            >
              {{ button.text }}
            </MazBtn>
          </template>
          <template v-else>
            <MazBtn
              v-if="cancelButtonData"
              v-bind="cancelButtonData"
              @click="rejectDialog(currentModal)"
            >
              <!--
                @slot cancel-text slot - Place your cancel text
              -->
              <slot name="cancel-text">
                {{ cancelButtonData.text }}
              </slot>
            </MazBtn>

            <MazBtn
              v-if="confirmButtonData"
              v-bind="confirmButtonData"
              @click="resolveDialog(currentModal)"
            >
              <!--
                @slot confirm-text slot - Place your confirm text
              -->
              <slot name="confirm-text">
                {{ confirmButtonData.text }}
              </slot>
            </MazBtn>
          </template>
        </div>
      </slot>
    </template>
  </MazDialog>
</template>
