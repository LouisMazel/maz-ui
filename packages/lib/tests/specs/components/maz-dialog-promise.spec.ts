import type { VueWrapper } from '@vue/test-utils'
import { shallowMount } from '@vue/test-utils'
import MazDialogPromise from '@components/MazDialogPromise.vue'
import { useMazDialogPromise } from '@components/MazDialogPromise/use-maz-dialog-promise'

describe('MazDialogPromise', () => {
  let wrapper: VueWrapper

  const { showDialogAndWaitChoice, removeDialogFromState } = useMazDialogPromise()

  beforeEach(() => {
    wrapper = shallowMount(MazDialogPromise, {
      props: {
        data: {
          title: 'Test Title',
          message: 'Test Message',
          confirmText: 'Confirm',
          cancelText: 'Cancel',
        },
        identifier: 'test',
      },
    })
  })

  afterEach(() => {
    removeDialogFromState('test')
  })

  test('Should match with the snapshot', async () => {
    showDialogAndWaitChoice('test')
    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders the correct title', () => {
    // @ts-ignore
    expect(wrapper.vm.currentModal).toBeUndefined()
    showDialogAndWaitChoice('test')
    // @ts-ignore
    expect(wrapper.vm.currentModal.id).toBeTruthy()
  })
})
