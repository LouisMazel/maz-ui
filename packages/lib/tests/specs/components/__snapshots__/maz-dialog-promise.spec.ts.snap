// Vitest Snapshot v1

exports[`MazDialogPromise > Should match with the snapshot 1`] = `<!--v-if-->`;
